File: go/internal/slhdsa/fors.go
package slhdsa

// FORS (Forest of Random Subsets) implementation per FIPS 205.

// FORSSKGen generates a FORS secret key value.
// FIPS 205 Algorithm 12.
func FORSSKGen(hs HashSuite, skSeed, pkSeed []byte, adrs *Address, idx int, n int) []byte {
	skAdrs := adrs.Copy()
	skAdrs.SetType(AddrForsPRF)
	skAdrs.SetKeyPairAddress(adrs.GetKeyPairAddress())
	skAdrs.SetTreeIndex(uint32(idx))
	return hs.PRF(pkSeed, skSeed, &skAdrs, n)
}

// FORSNode computes a node in a FORS tree.
// FIPS 205 Algorithm 13: fors_node(SK.seed, i, z, PK.seed, ADRS).
// adrs is never modified; local copies are used for all hash calls.
func FORSNode(hs HashSuite, skSeed, pkSeed []byte, i, z int, adrs *Address, p *Params) []byte {
	n := p.N
	kp := adrs.GetKeyPairAddress()

	if z == 0 {
		sk := FORSSKGen(hs, skSeed, pkSeed, adrs, i, n)
		nodeAdrs := adrs.Copy()
		nodeAdrs.SetType(AddrForsTree)
		nodeAdrs.SetKeyPairAddress(kp)
		nodeAdrs.SetTreeHeight(0)
		nodeAdrs.SetTreeIndex(uint32(i))
		return hs.F(pkSeed, &nodeAdrs, sk, n)
	}

	left := FORSNode(hs, skSeed, pkSeed, 2*i, z-1, adrs, p)
	right := FORSNode(hs, skSeed, pkSeed, 2*i+1, z-1, adrs, p)

	nodeAdrs := adrs.Copy()
	nodeAdrs.SetType(AddrForsTree)
	nodeAdrs.SetKeyPairAddress(kp)
	nodeAdrs.SetTreeHeight(uint32(z))
	nodeAdrs.SetTreeIndex(uint32(i))

	concat := make([]byte, 0, 2*n)
	concat = append(concat, left...)
	concat = append(concat, right...)
	return hs.H(pkSeed, &nodeAdrs, concat, n)
}

// FORSSign generates a FORS signature.
// FIPS 205 Algorithm 14: fors_sign(md, SK.seed, PK.seed, ADRS).
func FORSSign(hs HashSuite, md, skSeed, pkSeed []byte, adrs *Address, p *Params) []byte {
	n := p.N
	k := p.K
	a := p.A

	indices := base2b(md, a, k)

	sig := make([]byte, 0, k*(1+a)*n)

	for i := 0; i < k; i++ {
		idx := indices[i]
		absIdx := i*(1<<uint(a)) + idx // global leaf index in FORS forest

		// Secret key value
		sk := FORSSKGen(hs, skSeed, pkSeed, adrs, absIdx, n)
		sig = append(sig, sk...)

		// Authentication path: sibling at each height j
		// Global sibling index at height j = i*2^(a-j) + ((idx>>j)^1)
		for j := 0; j < a; j++ {
			globalSibling := i*(1<<uint(a-j)) + ((idx >> uint(j)) ^ 1)
			authNode := FORSNode(hs, skSeed, pkSeed, globalSibling, j, adrs, p)
			sig = append(sig, authNode...)
		}
	}

	return sig
}

// FORSPKFromSig computes the FORS public key from a signature.
// FIPS 205 Algorithm 15: fors_pkFromSig(SIG_FORS, md, PK.seed, ADRS).
func FORSPKFromSig(hs HashSuite, sigFORS, md, pkSeed []byte, adrs *Address, p *Params) []byte {
	n := p.N
	k := p.K
	a := p.A

	indices := base2b(md, a, k)

	roots := make([]byte, 0, k*n)
	offset := 0

	kp := adrs.GetKeyPairAddress()

	for i := 0; i < k; i++ {
		idx := indices[i]
		absIdx := i*(1<<uint(a)) + idx // global leaf index in FORS forest

		// Leaf from secret key
		sk := sigFORS[offset : offset+n]
		offset += n

		nodeAdrs := adrs.Copy()
		nodeAdrs.SetType(AddrForsTree)
		nodeAdrs.SetKeyPairAddress(kp)
		nodeAdrs.SetTreeHeight(0)
		nodeAdrs.SetTreeIndex(uint32(absIdx))
		node := hs.F(pkSeed, &nodeAdrs, sk, n)

		// Walk up authentication path
		for j := 0; j < a; j++ {
			authNode := sigFORS[offset : offset+n]
			offset += n

			nodeAdrs.SetTreeHeight(uint32(j + 1))
			nodeAdrs.SetTreeIndex(uint32(absIdx >> uint(j+1)))

			if (idx>>uint(j))%2 == 0 {
				concat := make([]byte, 0, 2*n)
				concat = append(concat, node...)
				concat = append(concat, authNode...)
				node = hs.H(pkSeed, &nodeAdrs, concat, n)
			} else {
				concat := make([]byte, 0, 2*n)
				concat = append(concat, authNode...)
				concat = append(concat, node...)
				node = hs.H(pkSeed, &nodeAdrs, concat, n)
			}
		}

		roots = append(roots, node...)
	}

	// Hash all roots together
	forsRootsAdrs := adrs.Copy()
	forsRootsAdrs.SetType(AddrForsRoots)
	forsRootsAdrs.SetKeyPairAddress(adrs.GetKeyPairAddress())
	return hs.Tl(pkSeed, &forsRootsAdrs, roots, n)
}

File: rust/slh-dsa/src/fors.rs
//! FORS — Forest of Random Subsets (FIPS 205, Algorithms 13-16).
//!
//! FORS is a few-time signature scheme used in SLH-DSA to sign the
//! message digest. It consists of `k` binary trees, each with `2^a`
//! leaves.

use alloc::vec::Vec;

use crate::address::{Address, AddressType};
use crate::hash::HashSuite;
use crate::utils::base_2b;

/// Generate a FORS secret key element (Algorithm 13).
///
/// Derives the secret value for leaf `idx` using PRF.
pub fn fors_skgen<H: HashSuite>(
    sk_seed: &[u8],
    pk_seed: &[u8],
    adrs: &Address,
    idx: u32,
    n: usize,
) -> Vec<u8> {
    let mut sk_adrs = adrs.copy();
    sk_adrs.set_type(AddressType::ForsPrf);
    sk_adrs.set_key_pair_address(adrs.get_key_pair_address());
    sk_adrs.set_tree_index(idx);
    H::prf(pk_seed, sk_seed, &sk_adrs, n)
}

/// Compute a FORS tree node via recursive treehash (Algorithm 14).
///
/// Returns the node at height `z` and horizontal index `i`.
/// The original `adrs` is never modified; local copies are used for all hash calls.
pub fn fors_node<H: HashSuite>(
    sk_seed: &[u8],
    i: u32,
    z: u32,
    pk_seed: &[u8],
    adrs: &Address,
    n: usize,
) -> Vec<u8> {
    let kp = adrs.get_key_pair_address();
    if z == 0 {
        // Leaf: hash the secret key element
        let sk = fors_skgen::<H>(sk_seed, pk_seed, adrs, i, n);
        let mut node_adrs = adrs.copy();
        node_adrs.set_type(AddressType::ForsTree);
        node_adrs.set_key_pair_address(kp);
        node_adrs.set_tree_height(0);
        node_adrs.set_tree_index(i);
        H::f(pk_seed, &node_adrs, &sk, n)
    } else {
        let left = fors_node::<H>(sk_seed, 2 * i, z - 1, pk_seed, adrs, n);
        let right = fors_node::<H>(sk_seed, 2 * i + 1, z - 1, pk_seed, adrs, n);
        let mut node_adrs = adrs.copy();
        node_adrs.set_type(AddressType::ForsTree);
        node_adrs.set_key_pair_address(kp);
        node_adrs.set_tree_height(z);
        node_adrs.set_tree_index(i);
        H::h(pk_seed, &node_adrs, &left, &right, n)
    }
}

/// FORS signature generation (Algorithm 15).
///
/// Signs the message digest `md` (which selects k leaf indices from k
/// FORS trees). Returns SIG_FORS = k * (sk_value + a auth nodes).
pub fn fors_sign<H: HashSuite>(
    md: &[u8],
    sk_seed: &[u8],
    pk_seed: &[u8],
    adrs: &Address,
    n: usize,
    k: usize,
    a: usize,
) -> Vec<u8> {
    let indices = base_2b(md, a as u32, k);
    let mut sig_fors = Vec::with_capacity(k * (1 + a) * n);

    for i in 0..k {
        let idx = indices[i];
        // Absolute leaf index within the FORS forest for this tree
        let abs_idx = (i as u32) * (1u32 << a) + idx;

        // Secret key value
        let sk = fors_skgen::<H>(sk_seed, pk_seed, adrs, abs_idx, n);
        sig_fors.extend_from_slice(&sk);

        // Authentication path for tree i
        let mut s = idx;
        for j in 0..(a as u32) {
            // Sibling node at height j
            let tree_offset = (i as u32) * (1u32 << (a as u32 - j));
            let node = fors_node::<H>(sk_seed, tree_offset + (s ^ 1), j, pk_seed, adrs, n);
            sig_fors.extend_from_slice(&node);
            s >>= 1;
        }
    }
    sig_fors
}

/// FORS public key recovery from signature (Algorithm 16).
///
/// Given a FORS signature and the message digest, recovers the FORS
/// public key by rebuilding each tree root and compressing them.
pub fn fors_pk_from_sig<H: HashSuite>(
    sig_fors: &[u8],
    md: &[u8],
    pk_seed: &[u8],
    adrs: &Address,
    n: usize,
    k: usize,
    a: usize,
) -> Vec<u8> {
    let indices = base_2b(md, a as u32, k);
    let mut roots = Vec::with_capacity(k * n);
    let kp = adrs.get_key_pair_address();

    let chunk_size = (1 + a) * n;

    for i in 0..k {
        let sig_chunk = &sig_fors[i * chunk_size..(i + 1) * chunk_size];
        let sk_val = &sig_chunk[..n];
        let auth = &sig_chunk[n..];

        let idx = indices[i];
        let abs_idx = (i as u32) * (1u32 << a) + idx;

        // Compute leaf from secret value
        let mut node_adrs = adrs.copy();
        node_adrs.set_type(AddressType::ForsTree);
        node_adrs.set_key_pair_address(kp);
        node_adrs.set_tree_height(0);
        node_adrs.set_tree_index(abs_idx);
        let mut node = H::f(pk_seed, &node_adrs, sk_val, n);

        // Walk up using auth path
        let mut s = idx;
        for j in 0..(a as u32) {
            let auth_j = &auth[(j as usize) * n..(j as usize + 1) * n];
            node_adrs.set_tree_height(j + 1);
            node_adrs.set_tree_index(abs_idx >> (j + 1));
            if s % 2 == 0 {
                node = H::h(pk_seed, &node_adrs, &node, auth_j, n);
            } else {
                node = H::h(pk_seed, &node_adrs, auth_j, &node, n);
            }
            s >>= 1;
        }
        roots.extend_from_slice(&node);
    }

    // Compress all k roots with T_k
    let mut fors_pk_adrs = adrs.copy();
    fors_pk_adrs.set_type(AddressType::ForsRoots);
    fors_pk_adrs.set_key_pair_address(kp);
    let blocks: Vec<&[u8]> = (0..k).map(|i| &roots[i * n..(i + 1) * n]).collect();
    H::t_l(pk_seed, &fors_pk_adrs, &blocks, n)
}
